node/hyperbahn/handler.js

```diff
@@ -338,3 +359,108 @@ function unadvertise(service) {
     var self = this;
     self.channel.topChannel.handler.removeServicePeer(service.serviceName, service.hostPort);
 };
+
+HyperbahnHandler.prototype.handleHostportsByService =
+function handleHostportsByService(self, req, arg2, arg3, cb) {
+    var serviceName = arg3.serviceName;
+    var exitNodes = self.egressNodes.exitsFor(serviceName);
+    var exitHosts = Object.keys(exitNodes);
+
+    var myHost = self.channel.hostPort;
+    if (exitHosts.indexOf(myHost) === -1) {
+        // Since Hyperbahn is fully connected to service hosts,
+        // randomly picking one suffices
+        var exit = Math.floor(Math.random() * exitHosts.length);
+        self.sendRelayAsThrift({
```

Comment:
two reasons: 1) generalizing this interface for moving other Hyperbahn endpoints to thrift. 2) retry is necessary, e.g., when k-value is being updated.